### DISCUSSION TITLE
React Refresh Does not work.

### DISCUSSION THREAD

**Discussion by gaurishhs** [vitejs/vite-plugin-react#76] [Q&A] [answered]
Hello everyone! I have a custom provider that provides all data provided by Vite from server-side vars using define in vite config.. Now whenever the page is refresh all the variables are undefined. What could be the issue?

```tsx
const [AppContextProvider, useAppContext] = constate(() => {
    const { logo, title, navLinks, sidebar, footer } = useMemo(() => {
        return __CODOC_APP_CONTEXT__;
    }, []);

    const routes = useMemo(() => {
        return asyncRoutes.find(item => item.path === '/')?.children ?? [];
    }, []);

    return {
        logo,
        title,
        routes,
        sidebar,
        navLinks,
        footer
    } as AppContext;
});
```

**Comment by Mhamad-Raad** [comment, marked as answer]
I am not sure if I understood the question correctly, but if you mean that when you refresh the page (which means the page loads again from the server) it is normal behavior that the context starts being empty again, I would suggest you use local-storage or anything of that sort if you want the data to stay or have default values, use it with useEffect, like the followings:

> useEffect(() => {
    if (!logo) setLogo(__CODOC_APP_CONTEXT__.logo);
    if (!title) setTitle(__CODOC_APP_CONTEXT__.title);
    if (!navLinks) setNavLinks(__CODOC_APP_CONTEXT__.navLinks);
    if (!sidebar) setSidebar(__CODOC_APP_CONTEXT__.sidebar);
    if (!footer) setFooter(__CODOC_APP_CONTEXT__.footer);
}, [logo, title, navLinks, sidebar, footer]);

It's also a good idea to check that the __CODOC_APP_CONTEXT__ is defined and only set it on client side.

> Copy code
let appContext;
if(typeof window !== 'undefined'){
   appContext = window.__CODOC_APP_CONTEXT__
}
